### PR TITLE
//scloud and cloud wand refactor

### PR DESCRIPTION
**Pull request by VorTechnix**
Pull request for #43

### CODE REVIEW

File: worldeditadditions/lib/selection/cloud.lua [hunk 1]
```diff
@@ -0,0 +1,48 @@
+--  ██████ ██       ██████  ██    ██ ██████
+-- ██      ██      ██    ██ ██    ██ ██   ██
+-- ██      ██      ██    ██ ██    ██ ██   ██
+-- ██      ██      ██    ██ ██    ██ ██   ██
+--  ██████ ███████  ██████   ██████  ██████
+worldeditadditions.add_pos = {}
+local wea = worldeditadditions
+function worldeditadditions.add_point(name, pos)
```
Comment (sbrl) [inline worldeditadditions/lib/selection/cloud.lua:8]: I feel the name of this function is too generic to be attached to the main `worldeditadditions` object. Perhaps it could be more descriptive? What about `sadditive_add` perhaps?

File: worldeditadditions_commands/commands/selectors/scloud.lua [hunk 2]
```diff
@@ -0,0 +1,35 @@
+-- ███████  ██████ ██       ██████  ██    ██ ██████
+-- ██      ██      ██      ██    ██ ██    ██ ██   ██
+-- ███████ ██      ██      ██    ██ ██    ██ ██   ██
+--      ██ ██      ██      ██    ██ ██    ██ ██   ██
+-- ███████  ██████ ███████  ██████   ██████  ██████
+local wea = worldeditadditions
+worldedit.register_command("scloud", {
+	params = "0-6/stop/reset",
+	description = "Set and add to WorldEdit region by punching put to six nodes that define the maximums of your target",
```
Comment (sbrl) [inline worldeditadditions_commands/commands/selectors/scloud.lua:9]: Typo
```suggestion
	description = "Set and add to WorldEdit region by punching up to six nodes that define the maximums of your target",
```

File: worldeditadditions_commands/commands/selectors/scloud.lua [hunk 3]
```diff
@@ -0,0 +1,35 @@
+-- ███████  ██████ ██       ██████  ██    ██ ██████
+-- ██      ██      ██      ██    ██ ██    ██ ██   ██
+-- ███████ ██      ██      ██    ██ ██    ██ ██   ██
+--      ██ ██      ██      ██    ██ ██    ██ ██   ██
+-- ███████  ██████ ███████  ██████   ██████  ██████
+local wea = worldeditadditions
+worldedit.register_command("scloud", {
+	params = "0-6/stop/reset",
```
Comment (sbrl) [inline worldeditadditions_commands/commands/selectors/scloud.lua:8]: This would be more consistent:
```suggestion
	params = "<0-6|stop|reset>",
```

File: worldeditadditions_commands/commands/selectors/scloud.lua [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+			end
+		elseif param == "stop" then
+			wea.add_pos[name] = nil
+			return true, "selection operation stopped"
+		elseif param == "reset" then
+			wea.add_pos[name] = nil
+			wea.clear_points(name)
+			return true, "selection cleared"
+		else
+			return false, (param == "" and "no input" or "invalid input: '"..param).."'! Allowed params are: 0-6/stop/reset"
```
Comment (sbrl) [inline worldeditadditions_commands/commands/selectors/scloud.lua:32]: Clarity could be aided here.
```suggestion
			return false, (param == "" and "no input" or "invalid input: '"..param).."'! Allowed params are: 0-6, stop, or reset"
```

File: worldeditadditions/lib/selection/cloud.lua [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		local volume_difference = volume_after - volume_before
+		
+		worldedit.marker_update(name)
+		worldedit.player_notify(name, "Expanded region by "..volume_difference.." nodes")
+	else
+		worldedit.player_notify(name, "Error: Too far away (try raising your maxdist with //farwand maxdist <number>)")
+		-- print("[set_pos1]", name, "nil")
+	end
+end
+function worldeditadditions.clear_points(name, pos)
```
Comment (sbrl) [inline worldeditadditions/lib/selection/cloud.lua:31]: Again, the name here could use some work. What about `sclear`?

File: worldeditadditions/lib/selection/cloud.lua [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+minetest.register_on_punchnode(function(pos, node, puncher)
+	local name = puncher:get_player_name()
+	if name ~= "" and wea.add_pos[name] ~= nil then
+		if wea.add_pos[name] > 0 then
+			wea.add_point(name,pos)
+			wea.add_pos[name] = wea.add_pos[name] - 1
+			worldedit.player_notify(name, "You have "..wea.add_pos[name].." nodes left to punch")
+		else wea.add_pos[name] = nil end
+	end
+end)
```
Comment (sbrl) [inline worldeditadditions/lib/selection/cloud.lua:48]: This belongs in `worldeditadditions_commands` I think, as it's a player-facing thingy like a chat command, not an API function.

File: worldeditadditions_commands/commands/selectors/scloud.lua [hunk 7]
```diff
@@ -0,0 +1,35 @@
+-- ███████  ██████ ██       ██████  ██    ██ ██████
+-- ██      ██      ██      ██    ██ ██    ██ ██   ██
+-- ███████ ██      ██      ██    ██ ██    ██ ██   ██
+--      ██ ██      ██      ██    ██ ██    ██ ██   ██
+-- ███████  ██████ ███████  ██████   ██████  ██████
```
Comment (sbrl) [inline worldeditadditions_commands/commands/selectors/scloud.lua:5]: Purely cosmetic lol - makes it look less like a 5 :P
```suggestion
--  ██████  ██████ ██       ██████  ██    ██ ██████
-- ██      ██      ██      ██    ██ ██    ██ ██   ██
-- ███████ ██      ██      ██    ██ ██    ██ ██   ██
--      ██ ██      ██      ██    ██ ██    ██ ██   ██
-- ██████   ██████ ███████  ██████   ██████  ██████
```

File: worldeditadditions_farwand/lib/cloudwand.lua [hunk 8]
```diff
@@ -1,52 +1,22 @@
-local function add_point(name, pos)
```
Comment (sbrl) [inline worldeditadditions_farwand/lib/cloudwand.lua:1]: Oh, good thinking. If we move this to the main `worldeditadditions` mod, then it can become common API logic across your new `//scloud` and my existing wand.

File: worldeditadditions_farwand/lib/cloudwand.lua [hunk 9]
```diff
@@ -1,52 +1,22 @@
...
-	worldedit.marker_update(name)
-	worldedit.set_pos[name] = nil
-	
-	worldedit.player_notify(name, "Region cleared")
-end
+local wea = worldeditadditions
 
 minetest.register_tool(":worldeditadditions:cloudwand", {
-	description = "WorldEditAdditions far-reaching point cloud wand",
+	description = "WorldEditAdditions far-reaching point cloud selection wand",
```
Comment (sbrl) [inline worldeditadditions_farwand/lib/cloudwand.lua:4]: My fault for not describing it better, but while we're here this description works better I think:
```suggestion
	description = "WorldEditAdditions far-reaching additive point cloud selection wand",
```

**Review by sbrl — CHANGES_REQUESTED** [review]
Just got a few suggestions. Smart thinking here with the code reuse!

File: worldeditadditions/lib/selection/cloud.lua (hunk 1 above)
Comment (VorTechnix) [reply worldeditadditions/lib/selection/cloud.lua:8]: `wea.selectors.add_point`?

File: worldeditadditions_commands/commands/selectors/scloud.lua (hunk 3 above)
Comment (VorTechnix) [reply worldeditadditions_commands/commands/selectors/scloud.lua:8]: world edit did it as
```
worldedit.register_command("p", {
	params = "set/set1/set2/get",
```
But if you really want it changed I'll commit this.

File: worldeditadditions/lib/selection/cloud.lua (hunk 5 above)
Comment (VorTechnix) [reply worldeditadditions/lib/selection/cloud.lua:31]: again let's just add `.selectors` to the name

File: worldeditadditions/lib/selection/cloud.lua (hunk 6 above)
Comment (VorTechnix) [reply worldeditadditions/lib/selection/cloud.lua:48]: I guess we could put it in scloud.lua

File: worldeditadditions_farwand/lib/cloudwand.lua (hunk 8 above)
Comment (VorTechnix) [reply worldeditadditions_farwand/lib/cloudwand.lua:1]: Yah it really makes things easier

File: worldeditadditions_farwand/lib/cloudwand.lua (hunk 9 above)
Comment (VorTechnix) [reply worldeditadditions_farwand/lib/cloudwand.lua:4]: Well now it's too wordy I think

Comment (VorTechnix) [reply worldeditadditions_farwand/lib/cloudwand.lua:4]: we could call it the selection addition wand or the additive selector wand.

File: worldeditadditions/lib/selection/cloud.lua (hunk 1 above)
Comment (sbrl) [reply worldeditadditions/lib/selection/cloud.lua:8]: `wea.selection.add_point`?

File: worldeditadditions_commands/commands/selectors/scloud.lua (hunk 3 above)
Comment (sbrl) [reply worldeditadditions_commands/commands/selectors/scloud.lua:8]: I see. I would if you don't mind, as it's more consistent with the rest of WorldEditAdditions (I use this syntax elsewhere).

File: worldeditadditions/lib/selection/cloud.lua (hunk 5 above)
Comment (sbrl) [reply worldeditadditions/lib/selection/cloud.lua:31]: Sure, that works. But I think `.selection` might be a slightly better name?

File: worldeditadditions/lib/selection/cloud.lua (hunk 6 above)
Comment (sbrl) [reply worldeditadditions/lib/selection/cloud.lua:48]: Yeah, that might be a better place for it.

File: worldeditadditions_farwand/lib/cloudwand.lua (hunk 9 above)
Comment (sbrl) [reply worldeditadditions_farwand/lib/cloudwand.lua:4]: Hrm, good point. What about this then?
```suggestion
	description = "WorldEditAdditions far-reaching additive selectior wand",
```

File: worldeditadditions/lib/selection/cloud.lua (hunk 5 above)
Comment (VorTechnix) [reply worldeditadditions/lib/selection/cloud.lua:31]: selection it is then

File: worldeditadditions/lib/selection/cloud.lua (hunk 6 above)
Comment (VorTechnix) [reply worldeditadditions/lib/selection/cloud.lua:48]: ok I'll change that

File: worldeditadditions/lib/selection/cloud.lua [hunk 10]
```diff
@@ -4,8 +4,7 @@
 -- ██      ██      ██    ██ ██    ██ ██   ██
 --  ██████ ███████  ██████   ██████  ██████
 worldeditadditions.add_pos = {}
-local wea = worldeditadditions
-function worldeditadditions.add_point(name, pos)
+function worldeditadditions.selection.add_point(name, pos)
```
Comment (sbrl) [inline worldeditadditions/lib/selection/cloud.lua:8]: You'll need to define `worldeditadditions.selection` like this if you haven't already:

```lua
worldeditadditions.selection = {}
```

Comment (VorTechnix) [reply worldeditadditions/lib/selection/cloud.lua:8]: this is now outdated

File: worldeditadditions_commands/commands/selectors/scloud.lua [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+			end
+		elseif param == "stop" then
+			wea.add_pos[name] = nil
+			return true, "selection operation stopped"
+		elseif param == "reset" then
+			wea.add_pos[name] = nil
+			wea.selection.clear_points(name)
+			return true, "selection cleared"
+		else
+			return false, (param == "" and "no input" or "invalid input: '"..param).."'! Allowed params are: 0-6, stop, or reset"
```
Comment (sbrl) [inline worldeditadditions_commands/commands/selectors/scloud.lua:42]: The quote is in the wrong place here
```suggestion
			return false, (param == "" and "no input" or "invalid input: '"..param.."'").."! Allowed params are: 0-6, stop, or reset"
```

**Review by sbrl — APPROVED** [review]
Thanks, @VorTechnix!
